### ISSUE TITLE
Vault token not passed?

### ISSUE THREAD

**Issue by carlpett** [hashicorp/consul-template#412]
I'm trying out consul-template along with vault, and just can't get my secrets out. I've set up a local consul and a dev vault, and added tried the following:

Add vault config to consul-template hcl:
```hcl
vault {
    address = "http://127.0.0.1:8200"
    token = "my-nice-token"
    ssl {
        enabled = false
    }
}
```

Sending it as an invokation var:

```hcl
VAULT_TOKEN="my-nice-token" consul-template ...
```

Exporting it to env:
```hcl
export VAULT_TOKEN="my-nice-token"
consul-template ...
```

In all cases, I get HTTP 400:
```hcl
URL: GET http://127.0.0.1:8200/v1/secret/stuff
Code: 400. Errors:

* missing client token
```

I did a quick search in the codebase, and I can't even find any references to the token? Is it all handled within some dependency?

**Comment by sethvargo** [comment]
Hi @carlpett 

What version of Consul Template are you running? Can you please run in debug mode and show the output here.

**Comment by carlpett** [comment]
Hi,
Latest release, 0.10, on Windows 8.1. Here's the log:

    $ ./consul-template -config consul-template-config.hcl -dry -once -log-level debug
    2015/09/30 09:43:35 [DEBUG] (config) loading configs from "consul-template-config.hcl"
    2015/09/30 09:43:35 [DEBUG] (logging) setting up logging
    2015/09/30 09:43:35 [DEBUG] (logging) config:

    {
      "name": "consul-template",
      "level": "debug",
      "syslog": false,
      "syslog_facility": "LOCAL0"
    }

    2015/09/30 09:43:35 [INFO] Consul Template v0.10.0
    2015/09/30 09:43:35 [INFO] (runner) creating new runner (dry: true, once: true)
    2015/09/30 09:43:35 [DEBUG] (runner) final config (tokens suppressed):

    {
      "path": "consul-template-config.hcl",
      "consul": "127.0.0.1:8500",
      "auth": {
        "enabled": false,
        "username": "",
        "password": ""
      },
      "vault": {
        "address": "http://127.0.0.1:8200",
        "ssl": {
          "enabled": false,
          "verify": true
        }
      },
      "ssl": {
        "enabled": false,
        "verify": true
      },
      "syslog": {
        "enabled": false,
        "facility": "LOCAL0"
      },
      "max_stale": 0,
      "templates": [
        {
          "source": "config/my-app.json.ctmpl",
          "destination": "config/my-app.json"
        }
      ],
      "retry": 5000000000,
      "wait": {
        "min": 0,
        "max": 0
      },
      "pid_file": "",
      "log_level": "debug"
    }

    2015/09/30 09:43:35 [INFO] (runner) creating consul/api client
    2015/09/30 09:43:35 [DEBUG] (runner) setting consul address to 127.0.0.1:8500
    2015/09/30 09:43:35 [INFO] (runner) creating vault/api client
    2015/09/30 09:43:35 [DEBUG] (runner) setting vault address to http://127.0.0.1:8200
    2015/09/30 09:43:35 [DEBUG] (runner) setting vault token
    2015/09/30 09:43:35 [INFO] (runner) creating Watcher
    2015/09/30 09:43:35 [INFO] (runner) starting
    2015/09/30 09:43:35 [DEBUG] (runner) running initial templates
    2015/09/30 09:43:35 [INFO] (runner) running
    2015/09/30 09:43:35 [DEBUG] (runner) checking template config/my-app.json.ctmpl
    2015/09/30 09:43:35 [INFO] (runner) was not watching 5 dependencies
    2015/09/30 09:43:35 [INFO] (watcher) adding "key(my-app/dev/server/port)"
    2015/09/30 09:43:35 [DEBUG] (watcher) "key(my-app/dev/server/port)" starting
    2015/09/30 09:43:35 [INFO] (watcher) adding "key(my-app/dev/database/adapter)"
    2015/09/30 09:43:35 [DEBUG] (watcher) "key(my-app/dev/database/adapter)" starting
    2015/09/30 09:43:35 [INFO] (watcher) adding "key(my-app/dev/database/host)"
    2015/09/30 09:43:35 [DEBUG] (watcher) "key(my-app/dev/database/host)" starting
    2015/09/30 09:43:35 [INFO] (watcher) adding "key(my-app/dev/database/database)"
    2015/09/30 09:43:35 [DEBUG] (watcher) "key(my-app/dev/database/database)" starting
    2015/09/30 09:43:35 [INFO] (watcher) adding "vault(secret/my-app/dev/database/credentials)"
    2015/09/30 09:43:35 [DEBUG] (watcher) "vault(secret/my-app/dev/database/credentials)" starting
    2015/09/30 09:43:35 [INFO] (runner) diffing and updating dependencies
    2015/09/30 09:43:35 [INFO] (runner) watching 5 dependencies
    2015/09/30 09:43:35 [DEBUG] (view) "key(my-app/dev/server/port)" starting fetch
    2015/09/30 09:43:35 [DEBUG] ("key(my-app/dev/server/port)") querying consul with &{Datacenter: AllowStale:false RequireConsistent:false WaitIndex:0 WaitTime:1m0s Token:}
    2015/09/30 09:43:35 [DEBUG] (view) "key(my-app/dev/database/adapter)" starting fetch
    2015/09/30 09:43:35 [DEBUG] ("key(my-app/dev/database/adapter)") querying consul with &{Datacenter: AllowStale:false RequireConsistent:false WaitIndex:0 WaitTime:1m0s Token:}
    2015/09/30 09:43:35 [DEBUG] (view) "key(my-app/dev/database/host)" starting fetch
    2015/09/30 09:43:35 [DEBUG] ("key(my-app/dev/database/host)") querying consul with &{Datacenter: AllowStale:false RequireConsistent:false WaitIndex:0 WaitTime:1m0s Token:}
    2015/09/30 09:43:35 [DEBUG] (view) "key(my-app/dev/database/database)" starting fetch
    2015/09/30 09:43:35 [DEBUG] ("key(my-app/dev/database/database)") querying consul with &{Datacenter: AllowStale:false RequireConsistent:false WaitIndex:0 WaitTime:1m0s Token:}
    2015/09/30 09:43:35 [DEBUG] (view) "vault(secret/my-app/dev/database/credentials)" starting fetch
    2015/09/30 09:43:35 [DEBUG] ("vault(secret/my-app/dev/database/credentials)") querying vault with &{AllowStale:false WaitIndex:0 WaitTime:1m0s}
    2015/09/30 09:43:35 [ERR] (view) "vault(secret/my-app/dev/database/credentials)" error reading from vault: Error making API request.

    URL: GET http://127.0.0.1:8200/v1/secret/my-app/dev/database/credentials
    Code: 400. Errors:

    * missing client token
    2015/09/30 09:43:35 [INFO] (view) "vault(secret/my-app/dev/database/credentials)" errored, retrying in 5s
    2015/09/30 09:43:35 [ERR] (runner) watcher reported error: error reading from vault: Error making API request.

    URL: GET http://127.0.0.1:8200/v1/secret/my-app/dev/database/credentials
    Code: 400. Errors:

    * missing client token
    2015/09/30 09:43:35 [INFO] (runner) running

**Comment by sethvargo** [comment]
Hi @carlpett 

And you have the token in the consul-template-config.hcl file? I just checked and this is working fine for me, so I think there's something less obvious with your environment going on.

**Comment by sethvargo** [comment]
And what version of Vault are you running?

**Comment by carlpett** [comment]
@sethvargo Tried multiple ways of supplying the token, when the log was taken it was in the hcl file, as well as exported to `VAULT_TOKEN`. Could it be Windows-related? I'm running it inside a cygwin bash prompt.

Vault version is 0.3 release. 

I noticed that in the debug output, the vault token is not part of the config dump, but I'm assuming that is for security reasons? And a few lines later, it says `2015/09/30 09:43:35 [DEBUG] (runner) setting vault token`.

**Comment by sethvargo** [comment]
@carlpett yea - the token is intentionally suppressed in the output for security reasons. I'll try this with Vault 0.3 in a bit.

**Comment by nbrownus** [comment]
Looks like the latest `consul-template` release was built against a version of the `vault` client that used cookie based auth, which doesn't work with `vault` v0.3. Rebuilding `consul-template` fixes the issue for me.

**Comment by carlpett** [comment]
@sethvargo Running with Vault 0.2 works fine, so something is related to 0.3

**Comment by itheiss** [comment]
Can confirm this.
@sethvargo: Any chance of getting a new consul-template release e.g. v0.10.1?

**Comment by sethvargo** [comment]
Soon

**Comment by sethvargo** [comment]
This will be fixed with the next release (which will be today or tomorrow at the latest) :smile:

**Comment by mhurne** [comment]
@sethvargo Was this in fact fixed? If so, in which release? I don't see mention of it in the release notes. Thanks!

**Comment by mhurne** [comment]
Also, @carlpett , were you able to confirm that the issue was resolved for you in some particular consul-template release?

**Comment by mhurne** [comment]
@sethvargo In general, it'd be nice to know which versions of consul-template are compatible with which versions of Vault...

**Comment by thalesfsp** [comment]
@mhurne passed one year, @sethvargo never mentioned the fix...

**Comment by nbrownus** [comment]
@thalesfsp The fix is to use a version of `vault` >= v0.3 and `consul-template` >= v0.11
